src/app/learnocaml_exercise_main.ml

```diff
@@ -93,6 +93,172 @@ let display_report exo report =
     (Format.asprintf "%a" Report.(output_html ~bare: true) report) ;
   grade
 
+let display_descr ex_meta =
+  let open Tyxml_js.Html5 in
+  let open Learnocaml_data.Exercise in
+  div ~a:[ a_class [ "descr" ] ] [
+    h2 ~a:[ a_class [ "learnocaml-exo-meta-category" ] ] [ pcdata ex_meta.Meta.title ] ;
+    p [ match ex_meta.Meta.short_description with
+        | None -> pcdata [%i"No description available."]
+        | Some text -> pcdata text ] ;
+  ]
+
+let display_stars ex_meta =
+  let open Tyxml_js.Html5 in
+  let open Learnocaml_data.Exercise in
+  div ~a:[ a_class [ "stars" ] ] [
+    let num = 5 * int_of_float (ex_meta.Meta.stars *. 2.) in
+    let num = max (min num 40) 0 in
+    let alt = Format.asprintf [%if"difficulty: %d / 40"] num in
+    let src = Format.asprintf "icons/stars_%02d.svg" num in
+    img ~alt ~src ()
+  ]
+
+let display_kind ex_meta =
+  let open Tyxml_js.Html5 in
+  let open Learnocaml_data.Exercise in
+  div ~a:[ a_class [ "length" ] ] [
+    match ex_meta.Meta.kind with
+    | Exercise.Meta.Project -> pcdata [%i"project"]
+    | Exercise.Meta.Problem -> pcdata [%i"problem"]
+    | Exercise.Meta.Exercise -> pcdata [%i"exercise"] ]
+
+(* Inefficient version: Reload the exercise description each time. *)
+let display_exercise_more content_id token id =
+  let open Tyxml_js.Html5 in
+  let open Learnocaml_data.Exercise in
+  let content = find_component content_id in
+  token >>= fun token ->
+  Server_caller.fetch_exercise token id >>= fun (meta, _, _) ->
```

Comment:
This call can be memoized: exercise description are static for a given running instance.